docs/design/features/objc-interop.md

```diff
@@ -0,0 +1,164 @@
+
+**Exceptions**
+
+The Objective-C implementation of exceptions is able to be correctly simulated through a series of well defined C functions provided by the Objective-C runtime. Details on this ABI aren't discussed here, but can be observed in various hand-written assembly code (e.g. [x86_64](https://github.com/xamarin/xamarin-macios/blob/main/runtime/trampolines-x86_64-objc_msgSend-post.inc)) in the Xamarin-macios code base.
+
+**Blocks**
+
+An Objective-C Block is conceptually a closure. The [Block ABI][blocks_abi] described by the clang compiler is used to inform interoperability with the .NET Platform. The Objective-C signature of a Block has an implied first argument for the Block itself. For example, a Block that takes an integer and returns an integer would have the follow C function signature: `int (*)(id, int)`.
+
+## Interaction between Objective-C and .NET types
+
+Creating an acceptable interaction model between Objective-C and .NET type instances hinges on addressing several issues.
+
+**Identity**
+
+The mapping between an Objective-C [`id`](https://developer.apple.com/documentation/objectivec/id) and a .NET `object` is of the upmost importance. Aside from issues around inefficient memory usage, the ability to know a .NET `object`'s true self is relied upon in many important scenarios (e.g. Key in a `Dictionary<K,V>`).
+
+Ensuring a robust mapping between these concepts can be handled more efficiently within a runtime implementation. This necessitates the ability for the consumer to register a pair (`id` and `object`) and whenever asked for one, the other can be returned.
+
+**Lifetime**
+
+Objective-C lifetime semantics are handled through [manual or automatic reference counting](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/MemoryMgmt.html). Reconciling this with the .NET Platform's Garbage Collector will require different solutions for each .NET runtime implementation.
+
+_Note_: In the following illustrations, a strong reference is depicted as a solid line (`===`) and a weak reference is depicted as a dashed line (`= = =`).
+
+When a .NET object enters an Objective-C environment, the Objective-C proxy is subject to reference counting and ensuring it extends the lifetime of the managed object it wraps. This can be accomplished in CoreCLR through use of the internal `HNDTYPE_REFCOUNTED` GC handle type. This handle, coupled with a reference count, can be used to transition a GC handle between a weak and strong reference.
+
+Creating an Objective-C proxy will also require overriding the built-in [`retain`](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1571946-retain) and [`release`](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1571957-release) methods provided by the Objective-C runtime.
```

Comment:
Each Objective-C type can implementing reference counting however it wishes, as long as the object is alive while the reference count is > 0, freed when reference count is = 0. This means that overriding `retain` and `release` won't give you as much information as you'd think: in particular it does not guarantee that you'll see all retains and releases (a type can increment/decrement the field holding the reference count outside of the `retain`/`release` methods), nor does it guarantee that `retain`/`release` will increment/decrement by one (a type can increment/decrement by two for instance). I've seen both of the scenarios, so they're not theoretical.

We override `retain`/`release` to:

* `retain`: detect when the reference count goes up, and in that case make sure the managed proxy isn't collected by the GC, because we know that there is some other native code that keeps the native object alive (technically we do this by switching from a weak GCHandle to a strong GCHandle).
* `release`: detect when the reference count goes from 2 to 1: do the opposite (switch from a strong GCHandle to a weak GCHandle).

Both of these are a best-effort thing, and we have other measures in place to cover the scenarios when these don't work:

* We missed a `retain`, and the GC tries to collect a managed object whose native object is retained by native code: in this case we have the toggle ref machinery, and the GC will ask us if a particular managed object can be freed, and we check the retain count.
* We missed a `release`, and we keep a strong reference to the managed object even though it's not needed: this would have to be detected by the user (through profiling for instance), and they'd have to manually call Dispose on the managed object to break the link between the native and the managed object. Fortunately I've never seen this happen.

This is ameliorated by the fact that we don't have to keep a managed object alive if it's only a wrapper around an existing native type, we can let the GC collect the object, and re-create it on demand later.

Ref: https://stackoverflow.com/a/13059140/183422

Reply:
This overriding would only occur if the most derived type is managed (i.e. defined in C#). All types originally defined in Objective-C would have their originally defined methods. This seems to be acceptable given the documentation on [`retain`](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1571946-retain) and [`release`](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1571957-release) that indicate one can implement reference counting however desired.

>You would implement this method only if you were defining your own reference-counting scheme. Such implementations must return self and should not invoke the inherited method by sending a retain message to super.

Given the above statement in the documentation, this model aligns with your statement about what one can expect but also implies the ownership of calling `dealloc` is left to the implementor of the `retain` and `release` since messages are not send to `super`.

This appears to be identical to the COM implementation of `AddRef`/`Release`. One facet that isn't in COM is support for an explicit reference count query (i.e. `retainCount`), but that would be trivial to also override. However the documentation for `retainCount` also states:

>Do not use this method.

So I am less concerned about needing to support that method.

Reply:
I believe I have addressed these issues in the latest updates coupled with the upcoming changes in the API proposal.